### DISCUSSION TITLE
Currency inputs

### DISCUSSION THREAD

**Discussion by MOA-TI** [livewire/flux#501] [Feature Requests]
First off, thank you for the fantastic work you’re doing with Flux!

I was wondering if there are any plans to introduce a currency input feature with options for custom decimal and thousand separators, currency symbols, right-aligned numbers, and more.

Additionally, would it be possible to mark ‘required’ field labels? This would allow them to be styled more easily using CSS.

These improvements would greatly simplify the development of web applications focused on management.

**Comment by jeffchown** [comment]
Hi @MOA-TI As Flux uses Alpine under the hood, have you tried:

```
<div>
    <flux:input label="Money Input" x-mask:dynamic="$money($input)" />
</div>
```

The documentation is here https://alpinejs.dev/plugins/mask#money-inputs

**Reply by lukas-staab** [reply]
Sadly x-mask:dynamic was not enough for me.

My goals: 
 * handle the number as `int` in php not as `float`, so you don't have resolution/rounding issues during float calculations 
 * Exchange the decimal seperator for `,` instead of `.` (x-mask does that on the frontend, but php side needs the `.` again)
 * format always with 2 visible decimals like `1.10` instead of `1.1` (which sadly x-mask=$money cannot do)

I ended after some extensive trial and error with: 

```blade
@props([
    'value' => 0,
])

<flux:input x-data="{ money : new Intl.NumberFormat('de-DE', {minimumFractionDigits:2}).format({{ $value }} / 100) }"
    class:input="text-right" x-model="money" x-mask:dynamic="$money($input, ',')"
    x-on:blur="$wire.set('{{ $attributes->wire('model')->value() }}', money.replace(',','.') * 100)"
/>

```
Usage: 
```blade 
<x-money-input wire:model="priceInCents" :value="$priceInCents" />
```


There are some parts which could be improved imho, namely: 
* right now it only updates on blur, i was not sure about the best way to make that dynamic
* how to access the value of wire:model from inside the blade component, right now I am doing it with a seperate variable `value`
* accepting / merging attributes which are given to `<x-money-input>`
* maybe initializing the `Intl.NumberFormat` object differently and manipulating the arguments via laravel config files or  blade component attributes
* changing mobile keyboards to number on focus (`input[type]=number` was not happy about exchanging the decimal seperator at my end)

But it works so far :)

The complexity for building it myself was higher then I anticipated (and its not even very polished yet). Therefore I would argue integrating it as a flux component would have some value.

**Comment by ju5t** [comment]
> I was wondering if there are any plans to introduce a currency input feature with options for custom decimal and thousand separators, currency symbols, right-aligned numbers, and more.

We use AutoNumeric for this, it requires a bit more effort but gives more flexibility. I don't think this should be added to Flux though, you can add it to a `flux:input` field so it requires no additional styling.

> Additionally, would it be possible to mark ‘required’ field labels? This would allow them to be styled more easily using CSS.

You can pass the `required` attribute to an input and use something like `required:border-red-500`.

**Reply by MOA-TI** [reply]
> > I was wondering if there are any plans to introduce a currency input feature with options for custom decimal and thousand separators, currency symbols, right-aligned numbers, and more.
> 
> We use AutoNumeric for this, it requires a bit more effort but gives more flexibility. I don't think this should be added to Flux though, you can add it to a `flux:input` field so it requires no additional styling.

Oh! I didn’t know about AutoNumeric.js! I’ll give it a try. Thanks for the heads up!!

> 
> > Additionally, would it be possible to mark ‘required’ field labels? This would allow them to be styled more easily using CSS.
> 
> You can pass the `required` attribute to an input and use something like `required:border-red-500`.

I’ve tried it, but unfortunately, only the input field is marked as ‘required’; neither the container (ui-field) nor the label (ui-label) are marked. Because of this, it’s not possible to apply CSS to change the label color or add a red * next to the label for the required fields. If the container had a data-* attribute (like data-flux-required), it would be much easier to style the required fields with CSS.

**Reply by MOA-TI** [reply]
> > I was wondering if there are any plans to introduce a currency input feature with options for custom decimal and thousand separators, currency symbols, right-aligned numbers, and more.
> 
> We use AutoNumeric for this, it requires a bit more effort but gives more flexibility. I don't think this should be added to Flux though, you can add it to a `flux:input` field so it requires no additional styling.

This is the code for my currency component:

```
@props([
    'symbol' => '€',
])

<div x-data="{ value: @entangle($attributes->wire('model')), autonumeric: undefined }"
     x-init="
        autonumeric = new AutoNumeric($refs.currency, value, {
           currencySymbol: ' {{ $symbol }}',
           currencySymbolPlacement: 's',
           decimalCharacter: ',',
           digitGroupSeparator: '.',
           formatOnPageLoad: true,
       });"
     class="input-right">
    <flux:input
        x-ref="currency"
        x-on:blur="value = autonumeric.getNumber()"
        type="text"
        {{ $attributes->whereDoesntStartWith('wire:model') }}
        x-model="value"/>
</div>
```

The CSS attribute "input-right" forces right alignment:

```
.input-right input {
   @apply text-right;
}

```

The component seems to work, but it only applies the formatting when the control is focused. Do you know how I can make it apply the formatting by default?

Thanks a lot!!

**Comment by boptom** [comment]
This is one I created to work with Alpine JS and also works within Flux:

https://boptom.github.io/alpinejs-number/
